R/diagnostic_test.R

```diff
@@ -0,0 +1,198 @@
+  results <- c(
+    check_proxy_settings(clean = clean, verbose = verbose),
+    check_github_pat(clean = clean, verbose = verbose),
+    check_renv_download_method(clean = clean, verbose = verbose)
+  )
+  return(results)
+}
+
+#' Check proxy settings
+#'
+#' @description
+#' This script checks for "bad" proxy settings. Prior to the pandemic, analysts
+#' in the DfE would need to add some proxy settings to their GitHub config.
+#' These settings now prevent Git from connecting to remote archives on GitHub
+#' and Azure DevOps if present, so this script identifies and (if clean=TRUE is
+#' set) removes them.
+#'
+#' @param proxy_setting_names Vector of proxy parameters to check for. Default:
+#' c("http.proxy", "https.proxy")
+#' @param clean Attempt to clean settings
+#' @param verbose Run in verbose mode
+#'
+#' @return List of problem proxy settings
+#' @export
+#'
+#' @examples
+#' check_proxy_settings()
+check_proxy_settings <- function(
+    proxy_setting_names = c("http.proxy", "https.proxy"),
+    clean = FALSE,
+    verbose = FALSE) {
+  git_config <- git2r::config()
+  proxy_config <- git_config |>
+    magrittr::extract2("global") |>
+    magrittr::extract(proxy_setting_names)
+  proxy_config <- purrr::keep(proxy_config, !is.na(names(proxy_config)))
+  if (any(!is.na(names(proxy_config)))) {
+    dfeR::toggle_message("Found proxy settings:", verbose = verbose)
+    paste(names(proxy_config), "=", proxy_config, collapse = "\n") |>
+      toggle_message(verbose = verbose)
+    if (clean) {
+      proxy_args <- proxy_config |>
+        lapply(function(list) {
+          NULL
+        })
+      rlang::inject(git2r::config(!!!proxy_args, global = TRUE))
+      message("FIXED: Git proxy settings have been cleared.")
+    } else {
+      message("FAIL: Git proxy setting have been left in place.")
+    }
+  } else {
+    message("PASS: No proxy settings found in your Git configuration.")
+    proxy_config <- as.list(rep("", length(proxy_setting_names))) |>
+      stats::setNames(proxy_setting_names)
+  }
+  return(proxy_config)
+}
+
+#' Check GITHUB_PAT setting
+#'
+#' @description
+#' If the GITHUB_PAT keyword is set, then it can cause issues with R installing
+#' packages from GitHub (usually with an error of "ERROR \[curl: (22) The
+#' requested URL returned error: 401\]"). This script checks whether the keyword
+#' is set and can then clear it (if clear=TRUE).
+#' The user will then need to identify where the "GITHUB_PAT" variable is being
+#' set from and remove it to permanently fix the issue.
+#'
+#' @inheritParams check_proxy_settings
+#'
+#' @return List object containing the github_pat keyword content.
+#' @export
+#'
+#' @examples
+#' check_github_pat()
+check_github_pat <- function(
+    clean = FALSE,
+    verbose = FALSE) {
+  github_pat <- Sys.getenv("GITHUB_PAT")
+  # Replace above to remove non alphanumeric characters when run on GitHub
+  # Actions
+  cat("==================================")
+  if (!is.na(github_pat)) {
```

Comment:
I think this is giving false positives, my GitHub PAT is unset but it's flagging as a fail. Should treat "" as the equivalent of NA?

![image](https://github.com/user-attachments/assets/83f2b231-a3fb-45f9-bc45-548272b7394b)

To illustrate it, even a made up variable I can guarantee has never been set on my machine returns as blank quotes rather than NA

![image](https://github.com/user-attachments/assets/5727d843-19af-45a1-9251-b5ae981d9508)